coral/src/services/api-client/index.test.ts

```diff
@@ -0,0 +1,173 @@
+import {
+  AbsolutePathname,
+  UnauthorizedError,
+} from "src/services/api-client/index";
+import { ClientError, ServerError } from "src/services/api-client/index";
+import api from "src/services/api-client";
+import { server } from "src/services/api-mocks/server";
+import { rest } from "msw";
+
+function apiUrl(path: string) {
+  return `${process.env.API_BASE_URL}${path}`;
+}
+
+type HTTPScenatio = {
+  functionName: string;
+  ok: () => Promise<unknown>;
+  htmlResponse: () => Promise<unknown>;
+  unauthorized: () => Promise<unknown>;
+  badRequest: () => Promise<unknown>;
+  internalError: () => Promise<unknown>;
+};
+
+describe.only("API client", () => {
+  const mockResponseData = { foo: "bar" };
+  beforeAll(() => {
+    server.listen();
+  });
+
+  beforeEach(() => {
+    server.use(
+      rest.all(apiUrl("/ok"), async (req, res, ctx) => {
+        return res.once(ctx.status(200), ctx.json(mockResponseData));
+      }),
+      rest.all(apiUrl("/okButHTML"), async (req, res, ctx) => {
+        return res.once(
+          ctx.status(200),
+          ctx.set("Content-Type", "text/html"),
+          ctx.text("<html></html>")
+        );
+      }),
+      rest.all(apiUrl("/unauthorized"), async (req, res, ctx) => {
+        return res.once(ctx.status(401), ctx.json(mockResponseData));
+      }),
+      rest.all(apiUrl("/clientError"), async (req, res, ctx) => {
+        return res.once(ctx.status(400), ctx.json(mockResponseData));
+      }),
+      rest.all(apiUrl("/serverError"), async (req, res, ctx) => {
+        return res.once(ctx.status(500), ctx.json(mockResponseData));
+      })
+    );
+  });
+
+  afterEach(() => {
+    server.resetHandlers();
+  });
+
+  function generateScenarioForMethodWithData(
+    name: string,
+    func: (
+      url: AbsolutePathname,
+      data: Record<string, string>
+    ) => Promise<unknown>
+  ): HTTPScenatio {
+    const data = { not: "relevant" };
+    return {
+      functionName: name,
+      ok: () => func("/ok", data),
+      htmlResponse: () => func("/okButHTML", data),
+      unauthorized: () => func("/unauthorized", data),
+      badRequest: () => func("/clientError", data),
+      internalError: () => func("/serverError", data),
+    };
+  }
+
+  function generateScenarioForMethod(
+    name: string,
+    func: (url: AbsolutePathname) => Promise<unknown>
+  ): HTTPScenatio {
+    return {
+      functionName: name,
+      ok: () => func("/ok"),
+      htmlResponse: () => func("/okButHTML"),
+      unauthorized: () => func("/unauthorized"),
+      badRequest: () => func("/clientError"),
+      internalError: () => func("/serverError"),
+    };
+  }
+
+  const GET: HTTPScenatio = generateScenarioForMethod("get", api.get);
+  const POST: HTTPScenatio = generateScenarioForMethodWithData(
+    "post",
+    api.post
+  );
+  const PUT: HTTPScenatio = generateScenarioForMethodWithData("puy", api.put);
```

Comment:
There's a typo with `puy` instead of `put` here. 

**Thought (nothing really to do)**
We could  add a stronger typing for this, but I'm not sure if it's worth it. 

I thought if we use `HTTPVerb` for that instead of string, we could also keep a `testCount` in the forEach loop and in the end check that the`testCount` equals the length of `HTTPVerb`,  but  while typing it out I realized that is a bit of an overkill, since the `HTTPVerb` length is nothing that is prone to changes 😅

Reply:
Not a big fan of exporting internal types / enums, so that they are accessed from tests, but I think here it makes sense. I will export `HTTPVerb` and use it here 👍